### PR TITLE
Bump SIP to 5.1.1 and PyQt to 5.14.1

### PR DESCRIPTION
**Pull request by Lctrs**

### CODE REVIEW

**Comment by Lctrs** [comment]
@bochecha I have one problem with the build of `PyQt`: it doesn't use the defined build flags in the `SDK` anymore. I tried various things to make it work but with no success. If you have any idea...

Otherwise, picard seems to work great.

**Comment by bochecha** [comment]
I'll try to have a look, when time permits.

Most likely the PyQt build system hard-codes a set of build flags, and either the way to override them changed, or they removed that possibility. So we need to look at their build system and figure out what they did, and what changed.

**Comment by Lctrs** [comment]
After a first glance at their generated Makefile, it seems that the build flags are indeed hardcoded now.

**Comment by Lctrs** [comment]
@bochecha New versions readded support for it o/

**Comment by Lctrs** [comment]
> It's a bit disappointing that moving to a more standard (for Python libs) buildsystem brings in more dependencies. disappointed
> 
> In addition, if we really want to move to installing these things with Pip, then we could just do it for PyQt5 as well, once this is fixed: [riverbankcomputing.com/pipermail/pyqt/2019-December/042397.html](https://www.riverbankcomputing.com/pipermail/pyqt/2019-December/042397.html)
> 
> So I'd recommend we wait until that's fixed before moving to installing everything with Pip?

According to the [documentation](https://www.riverbankcomputing.com/static/Docs/PyQt5/installation.html#building-and-installing-from-source), installing PyQt5 with `pip` is not recommended because it is not possible to configure the installation. `sip-install` is the recommended way to go.

File: org.musicbrainz.Picard.json [hunk 1]
```diff
@@ -140,14 +140,12 @@
...
-                "--disable-static"
-            ],
+            "buildsystem": "cmake-ninja",
             "sources": [
                 {
                     "type": "archive",
-                    "url": "http://ftp.musicbrainz.org/pub/musicbrainz/libdiscid/libdiscid-0.6.2.tar.gz", /* No HTTPS */
-                    "sha256": "f9e443ac4c0dd4819c2841fcc82169a46fb9a626352cdb9c7f65dd3624cd31b9"
+                    "url": "https://github.com/metabrainz/libdiscid/archive/v0.6.2.tar.gz",
+                    "sha256": "a9b73b030603ce707941a3aab4f46649fa5029025e7e2bfbbe0c3c93a86d7b20"
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Can you split this commit out to its own pull request, so we can merge it right away?

This way we can focus this merge request on the harder things related to the update (including the PyQt bump which seems complex)

File: flathub.json [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
-    "only-arches": ["x86_64", "i386"]
+    "only-arches": ["x86_64"]
```
Comment (bochecha) [inline flathub.json]: Can you split this commit out to its own pull request, so we can merge it right away?

This way we can focus this merge request on the harder things related to the update (including the PyQt bump which seems complex)

File: pyqt-build [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    --no-designer-plugin \
+    --no-qml-plugin \
+    --no-dbus-python \
+    --no-tools \
+    --concatenate 1 \
+    --debug \
+    --no-docstrings \
+    --qmake-setting QMAKE_CFLAGS="$CFLAGS" \
+    --qmake-setting QMAKE_CXXFLAGS="$CXXFLAGS" \
+    --qmake-setting QMAKE_LFLAGS="$LDFLAGS"
```
Comment (bochecha) [inline pyqt-build]: Could you reorder this file to make it as similar as possible to the old `pyqt-configure` script?

This would make it much easier to review the actual changes. :slightly_smiling_face:

Comment (Lctrs) [reply pyqt-build]: Is it good for you now ?

**Comment by Lctrs** [comment]
Now the question worth 1M$: how to remove the default flag `--ignore-installed` in from sip installation ? Any idea @bochecha ? I'm not familiar at all with python ecosystem.

File: pyqt-build (hunk 3 above)
Comment (bochecha) [reply pyqt-build]: Yes, this is much better, thank you. Thank you as well for splitting those other merge requests, this make sit all easier on my little brain. :slightly_smiling_face: 

I'm a bit disappointed that neither Git nor Github see that the new script is almost the same as the old one, and insists on treating them as a removal and an addition.

Nevertheless, now we can compare the two manually:

```diff
--- pyqt-configure	2020-02-10 15:02:33.256727180 +0100
+++ pyqt-build	2020-02-10 15:01:41.442990786 +0100
@@ -2,24 +2,25 @@
 
 PYTHON_VERSION=$(${PYTHON} -c 'import sys; print(sys.version[:3])')
 
-${PYTHON} configure.py \
-    --assume-shared \
-    --concatenate \
+sip-install \
+    --qt-shared \
+    --concatenate 1 \
     --confirm-license \
-    --destdir=${FLATPAK_DEST}/lib/python${PYTHON_VERSION}/site-packages \
+    --debug \
+    --verbose \
+    --build-dir=${FLATPAK_BUILDER_BUILDDIR}/tmp \
+    --target-dir=${FLATPAK_DEST}/lib/python${PYTHON_VERSION}/site-packages \
     --enable=QtCore \
+    --enable=QtDBus \
     --enable=QtGui \
     --enable=QtMultimedia \
     --enable=QtNetwork \
     --enable=QtWidgets \
+    --no-dbus-python \
     --no-designer-plugin \
-    --no-dist-info \
     --no-docstrings \
     --no-qml-plugin \
-    --no-qsci-api \
-    --no-sip-files \
-    --no-stubs \
     --no-tools \
-    QMAKE_CFLAGS_RELEASE="$CFLAGS" \
-    QMAKE_CXXFLAGS_RELEASE="$CXXFLAGS" \
-    QMAKE_LFLAGS_RELEASE="$LDFLAGS"
+    --qmake-setting QMAKE_CFLAGS="$CFLAGS" \
+    --qmake-setting QMAKE_CXXFLAGS="$CXXFLAGS" \
+    --qmake-setting QMAKE_LFLAGS="$LDFLAGS"
```

Below I'm going to try and review the above diff piece by piece, manually emulating the Github review UI.

```diff
-    --concatenate \
+    --concatenate 1 \
```

This surprised me, but looking at their old/new build systems, this change is entirely correct.

```diff
+    --debug \
+    --verbose \
```

This surprised me as well, because I thought `--debug` would enable verbose/debug output during the build, and I was wondering the difference with `--verbose`. It turns out the former is about generating builds with debuginfos (to be used by e.g `gdb`) so this is also entirely correct.

I wonder if our current builds just don't have debuginfos. :thinking:

```diff
+    --enable=QtDBus \
```

Why are you adding this? Is it something we're currently missing?

```diff
+    --no-dbus-python \
```

I'm guessing this is related to enabling `QtDBus`? This seems like [the right thing to do](https://dbus.freedesktop.org/doc/dbus-python/index.html#problems-and-alternatives).

```diff
-    --no-sip-files \
```

I seem to recall something about those "sip files" where not installing them would break something else's build later down, but I really can't remember what it was exactly. :disappointed:

**Comment by bochecha** [comment]
> how to remove the default flag --ignore-installed in from sip installation ?

I don't think this is the problem.

This flag is used when the sip build system tries to install its dependencies, namely `setuptools>=30.3` AND `wheel`.

The solution is not to make it install those deps without this flag, but instead to make sure it doesn't **not** try to install its deps at all. This is why we're passing `--no-deps` after all.

I'll try to have a look at what's going on soon.

File: pyqt-build (hunk 3 above)
Comment (Lctrs) [reply pyqt-build]: > I wonder if our current builds just don't have debuginfos. thinking

I sincerely have no idea.

> ```diff
> +    --enable=QtDBus \
> ```
> 
> Why are you adding this? Is it something we're currently missing?

Whoops. Leftover. It'll be required by the next Picard version.

**Comment by Lctrs** [comment]
> The solution is not to make it install those deps without this flag, but instead to make sure it doesn't **not** try to install its deps at all. This is why we're passing `--no-deps` after all.

Well, we did pass the `--no-deps` flag. Still it tries to install its dependencies.

**Comment by bochecha** [comment]
> Well, we did pass the --no-deps flag. Still it tries to install its dependencies.

**tld;dr:** Please add `--no-use-pep517` to every `pip` invocation.

---

Since you mentioned you're not very familiar with Python packaging, here's an explanation of what's going on.

Pip` is becoming the standard Python dependency installer, but that hasn't always been the case. There used to be others like easy_install at least, which came with the setuptools.

But there are multiple ways to package Python modules, for example distutils, setuptools, pbr, …If Pip wants to be the standard installer, it needs to be able to install modules no matter how they were packaged.

Let's take the example of the setuptools buildsystem as this is probably the most common. To package your module with setuptools, you create a `setup.py` file which contains the following:

```python
from setuptools import setup

setup(… some metadata, various packaging options …)
```

When installing your module, Pip will have to run that `setup.py` file, which will try to import the `setuptools` module. If you don't have that module already installed, Pip can't run the file and you get an ugly traceback.

That means Pip must have access to the `setuptools` module **before** it tries to run the `setup.py` script.

You can't add that with the rest of the dependencies in the `setup.py` script for obvious reasons. :grimacing: 

To fix that, the Python Packaging Authority introduced [PEP 517](https://www.python.org/dev/peps/pep-0517/) which describes a new standard for a `pyproject.toml` file where developers can put such "build dependencies" on things like the setuptools or other buildsystems.

In the case of the sip module, which fails to build in this merge request, this is the content of that file:

```toml
[build-system]
requires = ["setuptools >=30.3", "wheel"]
```

So now Pip knows that it needs to install the setuptools before running the `setup.py` script, which itself will contain the actual dependencies for the sip module.

We pass the `--no-deps` option, which only tells Pip not to install the module dependencies. But Pip still tries to install the build dependencies contained in `pyproject.toml`.

It turns out this is the expected behaviour, and the Pip developers [added the `--no-use-pep517` option](https://github.com/pypa/pip/pull/6370) to tell Pip not to look at the build dependencies contained in `pyproject.toml`.

So for the same reasons that we pass `--no-deps`, we should also pass `--no-use-pep517` to Pip, because it shouldn't deal with dependencies in a Flatpak context, we do that ourselves.

There still seems to be a bug that Pip doesn't see we already have `setuptools>=30.3` installed (the Sdk comes with version 41.2.0) and tries to install it. I don't know why it does that, but it doesn't matter much anyway: we still should pass `--no-use-pep517`.

Please add that option to all the Pip invocation in the manifest, and the build should succeed… unless a later module requires some other missing dependency. :slightly_smiling_face:

**Comment by Lctrs** [comment]
Thank you for the explanation. Much clearer now.

**Comment by Lctrs** [comment]
@bochecha If that's okay for you, let's merge this shit.

**Comment by bochecha** [comment]
> According to the [documentation](https://www.riverbankcomputing.com/static/Docs/PyQt5/installation.html#building-and-installing-from-source), installing PyQt5 with `pip` is not recommended because it is not possible to configure the installation. `sip-install` is the recommended way to go.

Right, they say:

> However using pip to install from the source package is not recommended as it is not possible to configure the installation or to easily diagnose any problems.

The ease of diagnosing issues seems to me like it comes mostly from habit: I'd have much less trouble diagnosing problems with a build system I'm more familiar with, and that's `pip` over `sip-install`. I assume this is the same for them, except they are more used to `sip-install` and understand better how to deal with it. :man_shrugging: 

However that probably implies that they'd fix issues faster with the `sip-install` build system if that's what they know best and recommend.

The part about configuring the installation is interesting. I'm guessing this is mostly about the `--enable=MODULE` arguments we pass. Losing that would indeed be annoying, because it would make the PyQt build take longer, and the resulting app bigger.

So all in all, let's keep it this way.

> @bochecha If that's okay for you, let's merge this shit.

Don't devalue your own work, this is not shit. It was hard labour, and it makes the Picard build better.

And let's merge it, yeah. :+1:

**Comment by Lctrs** [comment]
> Don't devalue your own work, this is not shit. It was hard labour, and it makes the Picard build better.

It was meant in a pseudo gangsta style. :smile:
